parquet-column/src/main/java/parquet/column/values/delta/DeltaBinaryPackingValuesWriterForInteger.java

```diff
@@ -0,0 +1,232 @@
+ * <p/>
+ * <pre>
+ *   {@code
+ *     delta-binary-packing: <page-header> <block>*
+ *     page-header := <block size in values> <number of miniblocks in a block> <total value count> <first value>
+ *     block := <min delta> <list of bitwidths of miniblocks> <miniblocks>
+ *
+ *     min delta : zig-zag var int encoded
+ *     bitWidthsOfMiniBlock : 1 byte little endian
+ *     blockSizeInValues,blockSizeInValues,totalValueCount,firstValue : unsigned varint
+ *   }
+ * </pre>
+ *
+ * The algorithm and format is inspired by D. Lemire's paper: http://lemire.me/blog/archives/2012/09/12/fast-integer-compression-decoding-billions-of-integers-per-second/
+ *
+ * @author Tianshuo Deng
+ */
+public class DeltaBinaryPackingValuesWriterForInteger {
+  /**
+   * max bitwidth for a mini block, it is used to allocate miniBlockByteBuffer which is
+   * reused between flushes.
+   */
+  public static final int MAX_BITWIDTH = 32;
+
+  /**
+   * stores blockSizeInValues, miniBlockNumInABlock and miniBlockSizeInValues
+   */
+  private final DeltaBinaryPackingConfig config;
+
+  /**
+   * bit width for each mini block, reused between flushes
+   */
+  private final int[] bitWidths;
+
+  private int totalValueCount = 0;
+
+  /**
+   * a pointer to deltaBlockBuffer indicating the end of deltaBlockBuffer
+   * the number of values in the deltaBlockBuffer that haven't flushed to baos
+   * it will be reset after each flush
+   */
+  private int deltaValuesToFlush = 0;
+
+  /**
+   * stores delta values starting from the 2nd value written(1st value is stored in header).
+   * It's reused between flushes
+   */
+  private int[] deltaBlockBuffer;
+
+  /**
+   * bytes buffer for a mini block, it is reused for each mini block.
+   * Therefore the size of biggest miniblock with bitwith of MAX_BITWITH is allocated
+   */
+  private byte[] miniBlockByteBuffer;
+
+  /**
+   * firstValue is written to the header of the page
+   */
+  private int firstValue = 0;
+
+  /**
+   * cache previous written value for calculating delta
+   */
+  private int previousValue = 0;
+
+  /**
+   * min delta is written to the beginning of each block.
+   * it's zig-zag encoded. The deltas stored in each block is actually the difference to min delta,
+   * therefore are all positive
+   * it will be reset after each flush
+   */
+  private int minDeltaInCurrentBlock = Integer.MAX_VALUE;
+
+  public DeltaBinaryPackingValuesWriterForInteger(final DeltaBinaryPackingConfig config) {
+    this.config = config;
+    bitWidths = new int[config.miniBlockNumInABlock];
+    deltaBlockBuffer = new int[config.blockSizeInValues];
+    miniBlockByteBuffer = new byte[config.miniBlockSizeInValues * MAX_BITWIDTH];
+  }
+
+  public void writeInteger(int v, final CapacityByteArrayOutputStream baos) {
+    totalValueCount++;
+
+    if (totalValueCount == 1) {
+      firstValue = v;
+      previousValue = firstValue;
+      return;
+    }
+
+    int delta = v - previousValue;//calculate delta
```

Comment:
That's a direct copy-paste from the previous version. I just duplicated the code and changed some INTs to LONGs. Unfortunately in Java AFAIK if the goal is high performance, duplicating the code is the best approach.

As for the overflow. My understanding is that it is meant to overflow. I am pretty sure that a formal proof can be given that this algorithm is correct.

I will try to give a more informal proof. According to the java specification, "The built-in integer operators do not indicate overflow or underflow in any way.". (http://stackoverflow.com/questions/5131131/what-happens-when-you-increment-an-integer-beyond-its-max-value). Therefore the algorithm practically works in the (mod INT32 signed) ring (http://en.wikipedia.org/wiki/Modular_arithmetic). A known properties of any modular ring is that:
a (mod INT32) + b (mod INT32) = a + b (mod INT32)
a (mod INT32) - b (mod INT32) = a - b (mod INT32)
The algorithm uses only additions and substractions, so by the properties above the decoded value is the same as the encoded, even though it might overflow a few times during the calculations. 

There is actually a unit test that checks exactly the case you mentioned - shouldReadMaxMinValue(). I have also added the INT64 equivalent shouldReadMaxMinValueLong() in this pull request.

One important note (that I was missing for quite some time). 
The INT32 version SHOULD NOT be written by the INT64 version encoder. The difference comes exactly from this overflow. INT_MIN - INT_MAX stored in int would still be a 32 bit integer, however stored in long would be 33 bit integer. Therefore given a list of 32 bit integers, encoded them with INT32 delta encoding might result in a different binary representation than encoding the same list of 32 bit integers with INT64 delta encoding. 

Nevertheless the INT64 decoding works for all possible cases. That's why I am using only one decoder - the INT64 decoder, and I am redirecting the INT32 decoder to it.